File: C#/CastleGame/Controller.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CastleGame
{
    class Controller
    {
        private string input = null;
        private string currentRoom = "Entrance Hall";
        private List<string> roomNames = new List<string>();
        //private List<string> items = new List<string>();
        private List<Room> rooms = new List<Room>();
        private List<Item> items = new List<Item>();
        private List<Item> bag = new List<Item>();
        private List<string> directions = new List<string>() { { "north" }, { "east" }, { "south" }, { "west" } };
        public Controller()
        {
            CreateRooms();
            foreach (string name in roomNames)
            {
                Console.WriteLine(name);
            }
            CreateItems();
            foreach (Item item in items)
            {
                Console.WriteLine($"{item.GetName()}: {item.GetRoom()}");
            }
            Console.WriteLine("Please enter your name"); //change to relevant text
            while (input != "quit")
            {
                Console.WriteLine($"You are currently in the {currentRoom}");
                Console.WriteLine(GetRoomDescription());
                Console.WriteLine(GetValidDirections());
                input = Console.ReadLine().ToString().ToLower();
                if(directions.Contains(input))
                {
                    Console.WriteLine($"You went {input}");
                    Move(input);
                }
                if(input != "quit" && directions.Contains(input) == false)
                {
                    Console.WriteLine($"{input} is not a valid direction");
                    Console.WriteLine(GetValidDirections());
                }
                
            }
            Console.WriteLine("Goodbye!");
        }

        private void CreateRooms()
        {
            roomNames.Add("Entrance Hall");
            roomNames.Add("Corridor");
            roomNames.Add("Kitchen");
            roomNames.Add("Armory");
            roomNames.Add("Ballroom");
            roomNames.Add("Smithy");
            roomNames.Add("Lookout");
            roomNames.Add("Dungeon");

            string Entrance_Hall_Desc = "You are in a large entrance hall with many paintings of previous owners. The faces seem to be obscured.";
            string Corridor_Desc = "You enter a long corridor. There is a scary suit of armor that always seems to be facing you, but without moving.";
            string Kitchen_Desc = "There is a pot bubbling on the stove in the kitchen. There is a particularly smelly piece of cheese that catches your eye.";
            string Armory_Desc = "You find many strange looking weapons in the armory. There are many different types of armor on stands, ready to be used if needed.";
            string Ballroom_Desc = "The ballroom is a large open expansive room. You can see a grandfather clock, a candlestick, a teapot and a teacup with a small chip.";
            string Smithy_Desc = "In the smithy, you find multiple tools for forging armor. There appears to be a small mouse running around.";
            string Lookout_Desc = "You find a large open window, perfect for spotting an attack. You can see a beautiful mountain range, with a river running towards the castle.";
            string Dungeon_Desc = "You travel down a spiral stone staircase to a dank and gloomy dungeon. You can see some light shining through the cracks of a locked door.";

            Dictionary<string, string> Entrance_Hall_Neighbours = new Dictionary<string, string>()
            {
                {"North","Corridor"}
            };
            Dictionary<string, string> Corridor_Neighbours = new Dictionary<string, string>()
            {
                {"North","Armory"},
                { "South","Entrance Hall"},
                { "West","Kitchen"}
            };
            Dictionary<string, string> Kitchen_Neighbours = new Dictionary<string, string>()
            {
                {"East","Corridor"}
            };
            Dictionary<string, string> Armory_Neighbours = new Dictionary<string, string>()
            {
                {"North","Ballroom"},
                {"South","Corridor"}
            };
            Dictionary<string, string> Ballroom_Neighbours = new Dictionary<string, string>()
            {
                {"North","Lookout"},
                {"East","Dungeon"}, 
                {"South","Armory"}, 
                {"West","Smithy"}
            };
            Dictionary<string, string> Smithy_Neighbours = new Dictionary<string, string>() 
            {
                {"East","Ballroom"}
            };
            Dictionary<string, string> Lookout_Neighbours = new Dictionary<string, string>() 
            { 
                {"South","Ballroom"}
            };
            Dictionary<string, string> Dungeon_Neighbours = new Dictionary<string, string>() 
            {
                {"West","Ballroom"}
            };

            Room Entrance_Hall = new Room("Entrance Hall", Entrance_Hall_Desc, Entrance_Hall_Neighbours);
            Room Corridor = new Room("Corridor", Corridor_Desc, Corridor_Neighbours);
            Room Kitchen = new Room("Kitchen", Kitchen_Desc,Kitchen_Neighbours );
            Room Armory = new Room("Armory", Armory_Desc,Armory_Neighbours);
            Room Ballroom = new Room("Ballroom", Ballroom_Desc, Ballroom_Neighbours);
            Room Smithy = new Room("Smithy", Smithy_Desc, Smithy_Neighbours);
            Room Lookout = new Room("Lookout", Lookout_Desc, Lookout_Neighbours);
            Room Dungeon = new Room("Dungeon", Dungeon_Desc, Dungeon_Neighbours);

            rooms.Add(Entrance_Hall);
            rooms.Add(Corridor);
            rooms.Add(Kitchen);
            rooms.Add(Armory);
            rooms.Add(Ballroom);
            rooms.Add(Smithy);
            rooms.Add(Lookout);
            rooms.Add(Dungeon);
        }

        private void CreateItems()
        {
            Item cheese = new Item("Cheese", "Kitchen");
            Item key = new Item("Key", "Smithy");

            items.Add(cheese);
            items.Add(key);
        }

        private string GetValidDirections()
        {
            string text = "No valid directions";
            foreach (Room room in rooms)
            {
                if(currentRoom == room.GetName())
                {
                    if (room.GetNeighbours().Count > 1)
                    {
                        text = "Available directions are ";
                    }
                    else
                    {
                        text = "Available direction is ";
                    }
                    foreach (KeyValuePair<string, string> entry in room.GetNeighbours())
                    {
                        //TODO Add a check for the last entry so the last two characters do not have to be removed later
                        text += $"{entry.Key}, ";
                    }
                }
            }
            //remove the last two characters from string (", ")
            return text[0..^2];
        }

        private void Move(String direction)
        {
            foreach (Room room in rooms)
            {
                if (currentRoom == room.GetName())
                {
                    Console.WriteLine("CurrentRoom");
                    foreach (KeyValuePair<string, string> entry in room.GetNeighbours())
                    {
                        if(direction == entry.Key.ToLower())
                        {
                            currentRoom = entry.Value;
                            Console.WriteLine($"You moved to the {entry.Value}");
                        }
                    }
                    break;
                }
            }
        }

        private string GetRoomDescription()
        {   string desc = "";
            foreach (Room room in rooms)
            {
                if(room.GetName() == currentRoom)
                {
                    desc = room.GetDesciption();
                }
            }
            return desc;
        }
    }
}

File: Java/src/Controller.java
import java.util.*;

public class Controller {
    private ArrayList<Room> rooms;
    private ArrayList<Item> items;
    private ArrayList<Item> bag;
    private String currentRoom;
    private ArrayList<String> keyWords;
    private String input;
    private boolean win;

    Controller(){
        currentRoom = "Entrance Hall";

        keyWords = new ArrayList<>();
        keyWords.add("go");
        keyWords.add("north");
        keyWords.add("east");
        keyWords.add("south");//
        keyWords.add("west");
        keyWords.add("take");
        keyWords.add("use");
        keyWords.add("bag");
        keyWords.add("help");
        keyWords.add("quit");

        bag =  new ArrayList<>();

        win = false;

        makeRooms();
        makeItems();
        startInput();
    }

    private void startInput() {
        Scanner scan = new Scanner(System.in);  // Create a Scanner object to read user input
        System.out.println("You enter a spooky castle");
        input = "";

        System.out.println(getDescription());
        System.out.println(getDirections());

        while(!input.equals("quit")){
            input = scan.nextLine().toLowerCase();  // Read user input and convert to lowercase
            String[] words = input.split(" ");
            if(keyWords.contains(words[0])){
                switch(words[0]){
                    case "go":
                        if(1 < words.length) {
                            move(words[1]);
                        }
                        else{
                            System.out.println("Go where?");
                        }
                        break;
                    case "take": // Take item
                        if(1 < words.length) {
                            System.out.println(takeItem(words[1]));
                        }
                        else{
                            System.out.println("Take what?");
                        }
                        break;
                    case "use": // Use item
                        if(1 < words.length){
                            System.out.println(useItem(words[1]));
                        }
                        else{
                            System.out.println("Use what?");
                        }
                        break;
                    case "bag": // Display inventory
                        System.out.println(getInventory());
                        break;
                    case "help": // Display keywords
                        System.out.println("You can use the following commands:");
                        System.out.println(getHelp());
                        break;
                    case "quit": // Quit the game
                        quit();
                        break;
                    default: // Move room
                        move(words[0]);
                }
            }
            else{
                System.out.println("Please enter a direction or command.");
                System.out.println("A list of commands can be found by entering 'help'");
            }
            if(win){
                quit();
            }
        }

    }

    private void makeRooms() {
        //room names
        String entranceName = "Entrance Hall";
        String corridorName = "Corridor";
        String kitchenName = "Kitchen";
        String armoryName = "Armory";
        String ballroomName = "Ballroom";
        String smithyName = "Smithy";
        String lookoutName = "Lookout";
        String dungeonName = "Dungeon";

        //room neighbours
        LinkedHashMap<String,String> entranceNeighbours = new LinkedHashMap<>();
        entranceNeighbours.put("North","Corridor");

        LinkedHashMap<String,String> corridorNeighbours = new LinkedHashMap<>();
        corridorNeighbours.put("North","Armory");
        corridorNeighbours.put("South","Entrance Hall");
        corridorNeighbours.put("West","Kitchen");

        LinkedHashMap<String,String> kitchenNeighbours = new LinkedHashMap<>();
        kitchenNeighbours.put("East","Corridor");

        LinkedHashMap<String,String> armoryNeighbours = new LinkedHashMap<>();
        armoryNeighbours.put("North","Ballroom");
        armoryNeighbours.put("South","Corridor");

        LinkedHashMap<String,String> ballroomNeighbours = new LinkedHashMap<>();
        ballroomNeighbours.put("North","Lookout");
        ballroomNeighbours.put("East","Dungeon");
        ballroomNeighbours.put("South","Armory");
        ballroomNeighbours.put("West","Smithy");

        LinkedHashMap<String,String> smithyNeighbours = new LinkedHashMap<>();
        smithyNeighbours.put("East","Ballroom");

        LinkedHashMap<String,String> lookoutNeighbours = new LinkedHashMap<>();
        lookoutNeighbours.put("South","Ballroom");

        LinkedHashMap<String,String> dungeonNeighbours = new LinkedHashMap<>();
        dungeonNeighbours.put("West","Ballroom");


        //room descriptions
        String entranceDesc = "You enter a large entrance hall with many paintings of previous owners. The faces seem to be obscured.";
        String corridorDesc = "You move into a long corridor. There is a scary suit of armor that always seems to be facing you, but without moving.";
        String kitchenDesc = "There is a pot bubbling on the stove in the kitchen. There is a particularly smelly piece of cheese that catches your eye.";
        String armoryDesc = "You find many strange looking weapons in the armory. There are many different types of armor on stands, ready to be used if needed.";
        String ballroomDesc = "The ballroom is a large open expansive room. You can see a grandfather clock, a candlestick, a teapot and a teacup with a small chip.";
        String smithyDesc = "In the smithy, you find multiple tools for forging armor. There appears to be a small mouse running around.";
        String lookoutDesc = "You find a large open window, perfect for spotting an attack. You can see a beautiful mountain range, with a river running towards the castle.";
        String dungeonDesc = "You travel down a spiral stone staircase to a dank and gloomy dungeon. You can see some light shining through the cracks of a locked door.";

        //create rooms
        Room entrance = new Room(entranceName,entranceNeighbours,entranceDesc);
        Room corridor = new Room(corridorName,corridorNeighbours,corridorDesc);
        Room kitchen = new Room(kitchenName,kitchenNeighbours,kitchenDesc);
        Room armory = new Room(armoryName,armoryNeighbours,armoryDesc);
        Room ballroom = new Room(ballroomName,ballroomNeighbours,ballroomDesc);
        Room smithy = new Room(smithyName,smithyNeighbours,smithyDesc);
        Room lookout = new Room(lookoutName,lookoutNeighbours,lookoutDesc);
        Room dungeon = new Room(dungeonName,dungeonNeighbours,dungeonDesc);

        //add rooms to ArrayList
        rooms = new ArrayList<>();

        rooms.add(entrance);
        rooms.add(corridor);
        rooms.add(kitchen);
        rooms.add(armory);
        rooms.add(ballroom);
        rooms.add(smithy);
        rooms.add(lookout);
        rooms.add(dungeon);
    }

    private void makeItems() {
        String cheesePickup = "You picked up the Cheese";
        String cheeseUse = "You give the cheese to the mouse. He grabs it and runs off knocking some items from a nearby shelf. You notice a key fall to the ground.";

        String keyPickup = "You picked up the Key";
        String keyUse = "Congratulations! You escaped from the spooky castle. Will you brave the castle once more and play again?";
        
        Item key = new Item("Cheese","Kitchen","Smithy",false,cheesePickup,cheeseUse);
        Item cheese = new Item("Key","Smithy", "Dungeon",false,keyPickup,keyUse);

        items = new ArrayList<>();
        items.add(key);
        items.add(cheese);
    }

    private void move(String direction){
        Room thisRoom = null;

        for(Room room : rooms){
            if(room.getName().equals(currentRoom)){
                thisRoom = room;
            }
        }

        assert thisRoom != null;
        HashMap<String, String> neighbours = thisRoom.getNeighbors();

        for(String dir : neighbours.keySet()){
            String room = neighbours.get(dir);

            if(dir.toLowerCase().equals(direction)){
                currentRoom = room;
            }
        }
        System.out.println(getDescription());
        System.out.println(getDirections());
    }

    private String getDescription(){
        String description = null;
        for(Room room: rooms){
            if(room.getName().equals(currentRoom)){
                description = room.getDescription();
            }
        }
        return description;
    }

    private String getDirections(){
        Room room = null;
        for(Room checkRoom : rooms){
            if(checkRoom.getName().equals(currentRoom)){
                room = checkRoom;
            }
        }
        assert room != null;
        HashMap<String,String> directions = room.getNeighbors();
        StringBuilder directionString = new StringBuilder("Available directions are ");
        for(String dir : directions.keySet()){
            directionString.append(dir).append(", ");
        }
        return directionString.substring(0, directionString.length() - 2);
    }

    private String takeItem(String takenItem){
        String itemString = "";
        boolean itemExists = false;
        for(Item item : items) {
            if (item.getName().toLowerCase().equals(takenItem)) {
                if(item.getRoom().equals(currentRoom)){
                    if (!item.getTaken()) {
                        bag.add(item);
                        item.setTaken(true);
                        itemExists = true;
                        itemString = item.getPickup();
                    } else {
                        itemString = "You have already picked this up";
                    }
                }
                else{
                    itemString = "That item doesn't exist!";
                }
            }
        }
        if(!itemExists){
            itemString = "That item doesn't exist!";
        }

        return itemString;
    }

    private String useItem(String usedItem){
        String useString = "";
        boolean inBag = false;
        for(Item item : bag){
            if(item.getName().toLowerCase().equals(usedItem)) {
                inBag = true;
                if (item.getUseRoom().equals(currentRoom)) {
                    useString = item.getUse();
                    bag.remove(item);
                    if(currentRoom.equals("Dungeon")){
                        win = true;
                        input = "quit";
                    }
                    break;
                }
                else{
                    useString = "This isn't the time to use that!";
                }
            }
        }
        if(!inBag){
            useString = "You don't have this item";
        }
        return useString;
    }

    private String getInventory(){
        StringBuilder inventoryString;
        if(bag.size() > 0) {
            inventoryString = new StringBuilder("You look into your bag. You find: ");
            for (Item item : bag) {
                inventoryString.append(item.getName()).append(", ");
            }
            inventoryString = new StringBuilder(inventoryString.substring(0, inventoryString.length() - 2));
        }
        else{
            inventoryString = new StringBuilder("Your bag is empty. Try looking for something to pickup.");
        }
        return inventoryString.toString();
    }

    private String getHelp(){
        StringBuilder commands = new StringBuilder();
        for(String word : keyWords){
            commands.append(word).append(", ");
        }
        return commands.substring(0, commands.length() - 2);
    }

    private void quit(){
        System.out.println("Thank you for playing! Goodbye!");
    }
}
